test/test_device_batch_memcpy.cu

```diff
@@ -0,0 +1,439 @@
+  size_t stats_src_offsets = sizeof(ByteOffsetT) * num_buffers;
+  size_t stats_dst_offsets = sizeof(ByteOffsetT) * num_buffers;
+  size_t stats_sizes       = sizeof(BufferSizeT) * num_buffers;
+  size_t stats_data_copied = 2 * num_total_bytes;
+
+  std::cout << "Min. buffer size: " << min_buffer_size << ", max. buffer size: " << max_buffer_size     //
+            << ", num_buffers: " << num_buffers                                                         //
+            << ", in_gen: " << ((input_gen == TestDataGen::RANDOM) ? "SHFL" : "CONSECUTIVE")            //
+            << ", out_gen: " << ((output_gen == TestDataGen::RANDOM) ? "SHFL" : "CONSECUTIVE")          //
+            << ", src size: " << stats_src_offsets << ", dst size: " << stats_dst_offsets               //
+            << ", sizes size: " << stats_sizes << ", cpy_data_size: " << stats_data_copied              //
+            << ", total: " << (stats_src_offsets + stats_dst_offsets + stats_sizes + stats_data_copied) //
+            << ", duration: " << duration                                                               //
+            << ", BW: "
+            << ((double)(stats_src_offsets + stats_dst_offsets + stats_sizes + stats_data_copied) / 1000000000.0) /
+                 (duration / 1000.0)
+            << "GB/s \n";
+#endif
+
+  for (ByteOffsetT i = 0; i < num_total_bytes; i++)
+  {
+    if (h_gpu_results[i] != h_out[i])
+    {
+      std::cout << "Mismatch at index " << i << ", CPU vs. GPU: " << static_cast<uint16_t>(h_gpu_results[i]) << ", "
+                << static_cast<uint16_t>(h_out[i]) << "\n";
+    }
+    AssertEquals(h_out[i], h_gpu_results[i]);
+  }
+
+  CubDebugExit(cudaFree(d_in));
+  CubDebugExit(cudaFree(d_out));
+  CubDebugExit(cudaFree(d_buffer_src_offsets));
+  CubDebugExit(cudaFree(d_buffer_dst_offsets));
+  CubDebugExit(cudaFree(d_buffer_sizes));
+  CubDebugExit(cudaFree(d_temp_storage));
+}
+
+int main(int argc, char **argv)
+{
+  CommandLineArgs args(argc, argv);
+
+  // Initialize device
+  CubDebugExit(args.DeviceInit());
+
+  // The most granular type being copied. Buffer's will be aligned and their size be an integer multiple of this type
+  using AtomicCopyT = uint8_t;
+
+  // Type used for indexing into the array of buffers
+  using BufferOffsetT = uint32_t;
+
+  // Type used for indexing into individual bytes of a buffer (large enough to cover the max buffer
+  using BufferSizeT = uint32_t;
+
+  // Type used for indexing into bytes over *all* the buffers' sizes
+  using ByteOffsetT = uint32_t;
+
+  // Each buffer's size will be random within this interval
+  std::vector<std::pair<BufferSizeT, BufferSizeT>> buffer_size_ranges = {
+    {1, 2},
+    {1, 32},
+    {1, 1024},
+    {1, 32 * 1024},
+  };
+
+  const BufferOffsetT target_copy_size = 16U << 20;
+  for (const auto &buffer_size_range : buffer_size_ranges)
+  {
+    BufferSizeT min_buffer_size = std::max(static_cast<BufferSizeT>(sizeof(AtomicCopyT)), buffer_size_range.first);
```

Comment:
Are we testing for empty buffers elsewhere?

Reply:
Good idea. I've now added a test case that comprises buffers that have `0 B` size.